engines/lab/dispman.cpp

```diff
@@ -0,0 +1,1022 @@
+	_vm->_anim->stopDiff();
+	loadPict(filename);
+	_vm->updateMusicAndEvents();
+
+	if (!_vm->_music->_loopSoundEffect)
+		_vm->_music->stopSoundEffect();
+
+	_dispBitMap->_bytesPerRow  = _screenWidth;
+	_dispBitMap->_drawOnScreen = (memoryBuffer == nullptr);
+	if (memoryBuffer)
+		_dispBitMap->_planes[0] = memoryBuffer;
+
+	_vm->_anim->readDiff(_curBitmap, playOnce, onlyDiffData);
+}
+
+void DisplayMan::freePict() {
+	delete _curBitmap;
+	_curBitmap = nullptr;
+}
+
+Common::String DisplayMan::getWord(const char *mainBuffer) {
+	Common::String result;
+
+	for (int i = 0; mainBuffer[i] && (mainBuffer[i] != ' ') && (mainBuffer[i] != '\n'); i++)
+		result += mainBuffer[i];
+
+	return result;
+}
+
+Common::String DisplayMan::getLine(TextFont *tf, const char **mainBuffer, uint16 lineWidth) {
+	uint16 curWidth = 0;
+	Common::String result;
+	bool doit = true;
+
+	lineWidth += textLength(tf, " ");
+
+	while ((*mainBuffer)[0] && doit) {
+		Common::String wordBuffer = getWord(*mainBuffer) + " ";
+
+		if ((curWidth + textLength(tf, wordBuffer)) <= lineWidth) {
+			result += wordBuffer;
+			(*mainBuffer) += wordBuffer.size() - 1;
+
+			if ((*mainBuffer)[0] == '\n')
+				doit = false;
+
+			if ((*mainBuffer)[0])
+				(*mainBuffer)++;
+
+			curWidth = textLength(tf, result);
+		} else
+			doit = false;
+	}
+
+	return result;
+}
+
+int DisplayMan::flowText(TextFont *font, int16 spacing, byte penColor, byte backPen,
+			bool fillBack, bool centerh, bool centerv, bool output, Common::Rect textRect, const char *str) {
+	if (fillBack) {
+		setPen(backPen);
+		rectFill(textRect);
+	}
+
+	if (!str)
+		return 0;
+
+	setPen(penColor);
+
+	TextFont *msgFont = font;
+	uint16 fontHeight = textHeight(msgFont) + spacing;
+	uint16 numLines   = (textRect.height() + 1) / fontHeight;
+	uint16 width      = textRect.width() + 1;
+	uint16 y          = textRect.top;
+	Common::String lineBuffer;
+
+	if (centerv && output) {
+		const char *temp = str;
+		uint16 actlines = 0;
+
+		while (temp[0]) {
+			lineBuffer = getLine(msgFont, &temp, width);
+			actlines++;
+		}
+
+		if (actlines <= numLines)
+			y += ((textRect.height() + 1) - (actlines * fontHeight)) / 2;
+	}
+
+	int len = 0;
+	while (numLines && str[0]) {
+		lineBuffer = getLine(msgFont, &str, width);
+
+		uint16 x = textRect.left;
+		len += lineBuffer.size();
+
+		if (centerh)
+			x += (width - textLength(msgFont, lineBuffer)) / 2;
+
+		if (output)
+			drawText(msgFont, x, y, penColor, lineBuffer);
+
+		numLines--;
+		y += fontHeight;
+	}
+
+	len--;
+
+	return len;
+}
+
+int DisplayMan::flowTextToMem(Image *destIm, TextFont *font, int16 spacing, byte penColor,
+			byte backPen, bool fillBack, bool centerh, bool centerv, bool output, Common::Rect textRect,
+			const char *str) {
+	byte *saveDisplayBuffer = _currentDisplayBuffer;
+	uint32 bytesPerPage = _screenBytesPerPage;
+
+	_currentDisplayBuffer = destIm->_imageData;
+	_screenBytesPerPage = (uint32)destIm->_width * (int32)destIm->_height;
```

Comment:
This just looks odd to cast the operands to different types. In general I saw a lot of arithmetics using casts, making me wonder about the types used in general. So, if someone has too much time on their hands, one could try to clean that up a lot.